api/db/base.js

```diff
@@ -149,101 +158,114 @@ const classExtension = (orm, models) => ({
-        logger.silly(null, raw);
-
-        const insert = this.forge({
-          ...this.pickUpdateable(raw),
-          ...parentID
-        });
-        await insert.save(null, { transacting });
+      // TODO [GW] Should we add an "order" to all of the tables that
+      // represent lists in their parent tables?  Instead of relying
+      // on IDs.  Currently there's no way to re-order things except
+      // to delete and re-add them in the order you want them.  It
+      // would be pretty trivial to dynamically add an "order" property
+      // based on an item's array index here, and then make all of those
+      // things automatically sort by order on fetch.
+
+      const saving = [];
+      const synchronizing = dataArray.filter(d => d.id).map(async raw => {
+        const existing = all.find(row => row.get('id') === raw.id);
+        if (existing) {
+          logger.silly(`${this.modelName} | updating id=${raw.id}`);
 
-        synchronizing.push(
-          syncChildren(
+          existing.set(this.pickUpdateable(raw));
+          saving.push(existing.save(null, { transacting }));
+
+          return syncChildren(
             raw,
             childModels,
             this.models,
-            { [this.foreignKey]: insert.get('id') },
+            { [this.foreignKey]: existing.get('id') },
+            deleteOthers,
             transacting,
             this.modelName
-          )
-        );
+          );
+        }
+        return false;
+      });
+
+      let inserting = Promise.resolve();
+      dataArray.filter(d => !d.id).map(async raw => {
+        // Chain these promises so they execute in the same
+        // order as the array.  This way the IDs are also in
+        // order, and we can sort things by ID.
+        inserting = inserting.then(async () => {
+          logger.silly(`${this.modelName} | inserting new`);
+          logger.silly(null, raw);
+
+          const insert = this.forge({
+            ...this.pickUpdateable(raw),
+            ...parentID
+          });
+          await insert.save(null, { transacting });
+
+          synchronizing.push(
+            syncChildren(
+              raw,
+              childModels,
+              this.models,
+              { [this.foreignKey]: insert.get('id') },
+              deleteOthers,
+              transacting,
+              this.modelName
+            )
+          );
+        });
       });
-    });
 
-    // Wait for the updates, inserts, and child
-    // synchronizations to finish, then we're done.
-    await Promise.all(saving);
-    await inserting;
-    await Promise.all(synchronizing);
+      // Wait for the updates, inserts, and child
+      // synchronizations to finish, then we're done.
+      await Promise.all(saving);
+      await inserting;
+      await Promise.all(synchronizing);
+    };
+
+    if (transaction) {
+      await exec(transaction);
+    } else {
+      await orm.transaction(transacting => exec(transacting));
+    }
```

Comment:
If you call `synchronize` on the model class outside of a transaction, this code will wrap it up in a transaction.  A whole lot of green lines just to represent indentation changes.  😛